aws-lc-rs/src/tls_prf.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! TLS 1.2 PRF API's for usage in [RFC 5246](https://www.rfc-editor.org/rfc/rfc5246) and [RFC 7627](https://www.rfc-editor.org/rfc/rfc7627).
+
+use crate::{
+    digest::match_digest_type, digest::AlgorithmID, error::Unspecified, fips::indicator_check,
+};
+use core::ptr::null;
+
+use aws_lc::CRYPTO_tls1_prf;
+
+/// The TLS PRF P_hash Algorithm
+pub struct Algorithm(AlgorithmID);
+
+/// SHA-256 P_hash algorithm
+pub const P_SHA256: &Algorithm = &Algorithm(AlgorithmID::SHA256);
+
+/// SHA-384 P_hash algorithm
+pub const P_SHA384: &Algorithm = &Algorithm(AlgorithmID::SHA384);
+
+/// SHA-512 P_hash algorithm
+pub const P_SHA512: &Algorithm = &Algorithm(AlgorithmID::SHA512);
+
+/// Encapsulates a PRF algorithm and secret bytes to be used to derive output.
+pub struct Secret<'a> {
```

Comment:
* I'm not crazy about having a struct with a lifetime; reuse requires this struct to tag along with whatever struct actually owns the "secret" value it's referencing. The alternatives are to either make the struct generic or put the secret in a Box, which also have downsides. Could this be a trait with a default implementation? Are there other alternatives?
* The name "Secret" is super-generic for a type with only one real use, but when prefixed with its owning module  (`tls_prf::Secret`) the purpose is clear, so maybe this is expected in Rust. I found guidance in the [Rust API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#iterator-type-names-match-the-methods-that-produce-them-c-iter-ty) that seems to support the use of generic type name: "*These type names make the most sense when prefixed with their owning module, for example vec::IntoIter.*".
 * Should the label be persisted in this struct as well?  How is this commonly used?

Reply:
> I'm not crazy about having a struct with a lifetime ... The alternatives are to either make the struct generic or put the secret in a Box, which also have downsides.

I opted for the lifetime as the secret (called the pre_master_secret in [RFC 7627]) size is dependent on the key exchange protocol, thus didn't want to allocate the worst case buffer size. A box was the alternative choice that I had in mind.

> Should the label be persisted in this struct as well? How is this commonly used?

From all indications these are really meant to be used in a one-shot manner in the protocols either using a label of "master secret", the Record Protocol related keys and IVs used where the labels are one or more of ("key expansion", "client write key", and "server write key", "IV block") depending on the whether TLS 1.0, 1.1, 1.2, and the "extended master secret" extension. So the flow is that a set of inputs (secret, label, seed) will be used to derive some variable length of output key material (secret') that can then subsequently be used to derive more material via (secret', label', seed'). It's vaguely similar to the HKDF module in some sense, so I might look at that to see if we could map similar concepts to roughly match the design, but keep it intuitive.

TLS 1.0: [RFC 2246](https://www.rfc-editor.org/rfc/rfc2246)
TLS 1.1: [RFC 4346](https://www.rfc-editor.org/rfc/rfc4346)
TLS 1.2: [RFC 5246](https://www.rfc-editor.org/rfc/rfc5246)
Extended Master Secret Extension: [RFC 7627](https://www.rfc-editor.org/rfc/rfc7627)